aspnetcore/host-and-deploy/health-checks.md

```diff
@@ -0,0 +1,359 @@
+```
+
+### Kubernetes example
+
+Using separate readiness and liveness checks is useful in an environment such as [Kubernetes](https://kubernetes.io/). In Kubernetes, an app might be required to perform time-consuming startup work before accepting requests, such as a test of the underlying database availability. Using separate checks allows the orchestrator to distinguish whether the app is functioning but not yet ready or if the app has failed to start. For more information on readiness and liveness probes in Kubernetes, see [Configure Liveness and Readiness Probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/) in the Kubernetes documentation.
+
+The following example demonstrates a Kubernetes readiness probe configuration:
+
+```
+spec:
+  template:
+  spec:
+    readinessProbe:
+      # an http probe
+      httpGet:
+        path: /health/ready
+        port: 80
+      # length of time to wait for a pod to initialize
+      # after pod startup, before applying health checking
+      initialDelaySeconds: 30
+      timeoutSeconds: 1
+    ports:
+      - containerPort: 80
+```
+
+## Metric-based probe (memory) with a custom response writer
+
+The sample app demonstrates a memory health check with a custom response writer.
+
+`MemoryHealthCheck` reports a degraded status if the app is using more than 1 GB of memory. The `HealthCheckResult` includes Garbage Collector (GC) information for the app (*MemoryHealthCheck.cs*):
+
+[!code-csharp[](health-checks/samples/2.x/HealthChecksSample/MemoryHealthCheck.cs?name=snippet1&highlight=9-16,19-27)]
+
+Register health check services with `AddHealthChecks` in `Startup.ConfigureServices`. Instead of enabling the health check by passing it to `AddCheck`, the `MemoryHealthCheck` is registered as a service. All `IHealthCheck` registered services are available to the health check services and middleware. We recommend registering health check services as Singleton services.
+
+*CustomWriterStartup.cs*:
+
+[!code-csharp[](health-checks/samples/2.x/HealthChecksSample/CustomWriterStartup.cs?name=snippet_ConfigureServices&highlight=4)]
+
+Call Health Check Middleware in the app processing pipeline in `Startup.Configure`. A `WriteResponse` delegate is provided to the `ResponseWriter` property to output a custom JSON response when the health check executes:
+
+[!code-csharp[](health-checks/samples/2.x/HealthChecksSample/CustomWriterStartup.cs?name=snippet_Configure&highlight=6)]
+
+The `WriteResponse` method formats the `CompositeHealthCheckResult` into a JSON object and yields JSON output for the health check response:
+
+[!code-csharp[](health-checks/samples/2.x/HealthChecksSample/CustomWriterStartup.cs?name=snippet_WriteResponse)]
+
+To run the metric-based probe with custom response writer output using the sample app, execute the following command from the project's folder in a command shell:
+
+```console
+dotnet run --scenario writer
+```
```

Comment:
This is another thing that the new BeatPulse libraries will provide. Our guidance will be to use those when available.